### PR TITLE
[TokenIndexer] Setup service

### PR DESCRIPTION
**Pull request by jeyldii**

### CODE REVIEW

File: api/registry.go [hunk 1]
```diff
@@ -108,8 +109,14 @@ func RegisterBasicAPI(router gin.IRouter) {
...
 
-func RegisterTokensIndexAPI(router gin.IRouter, instance tokensearcher.Instance) {
+func RegisterTokensSearcherAPI(router gin.IRouter, instance tokensearcher.Instance) {
 	router.POST("/v3/tokens", func(c *gin.Context) {
 		endpoint.GetTokensByAddressIndexer(c, instance)
 	})
 }
+
+func RegisterTokensIndexAPI(router gin.IRouter, instance tokenindexer.Instance) {
+	router.POST("/v3/tokens/new", func(c *gin.Context) {
```
Comment (vikmeup) [inline api/registry.go:119]: Should this be GET request?

File: db/asset.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			err := tx.Clauses(clause.OnConflict{DoNothing: true}).Create(&na).Error
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func assetsBatch(values []models.Asset, sizeUint uint) [][]models.Asset {
```
Comment (vikmeup) [inline db/asset.go:57]: Can you move this into local lib? I think this is used somewhere else already

File: db/asset.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		}
+		result[i] = values[lo:hi:hi]
+		lo, hi = hi, hi+size
+	}
+	return result
+}
+
+func (i *Instance) GetAssetsByIDs(ids []string, ctx context.Context) ([]models.Asset, error) {
+	db := i.Gorm.WithContext(ctx)
+	if len(ids) == 0 {
```
Comment (vikmeup) [inline db/asset.go:74]: Can you delete this check or remove it out?

File: pkg/blockatlas/tx.go [hunk 4]
```diff
@@ -452,3 +456,97 @@ func GetEthereumTokenTypeByIndex(coinIndex uint) TokenType {
 	}
 	return tokenType
 }
+
+func (t Tx) AssetModel() (models.Asset, bool) {
+	var asset models.Asset
+	switch t.Meta.(type) {
```
Comment (vikmeup) [inline pkg/blockatlas/tx.go:462]: I have seen very similar code with memo implementation for spam, can it be reused?

File: platform/tron/token_test.go [hunk 5]
```diff
@@ -102,7 +102,7 @@ func createMockedAPI() http.Handler {
...
...ls":18,"token_id":"TCFLL5dx5ZJdKnWuesXxi1VPwjLVmWZZy9","coin":195,"type":"TRC20"},{"name":"Enme Token","symbol":"EME","decimals":6,"token_id":"TCRhVHPv6efvXgogNMhiunAMXFKcMmv2pF","coin":195,"type":"TRC20"},{"name":"BeeHive","symbol":"BEE","decimals":8,"token_id":"TG7Z1ptC7nRkaniDVRhHSyLycaroaS5PdK","coin":195,"type":"TRC20"},{"name":"Mono Token","symbol":"MONO","decimals":18,"token_id":"TLKyLt4MXuvvdFvUUc3Zma4rZyj2t87Mak","coin":195,"type":"TRC20"},{"name":"WINK","symbol":"WIN","decimals":6,"token_id":"TLa2f6VPqDgRE67v1736s7bJ8Ray5wYjU7","coin":195,"type":"TRC20"},{"name":"PYRO Network","symbol":"PYRO","decimals":6,"token_id":"TMCMPzmosnQ8UAYW1zcBwjLTxDq8ce4Y5e","coin":195,"type":"TRC20"},{"name":"NoleCoin","symbol":"NOLE","decimals":6,"token_id":"TPt8DTDBZYfJ9fuyRjdWJr4PP68tRfptLG","coin":195,"type":"TRC20"},{"name":"Tether USD","symbol":"USDT","decimals":6,"token_id":"TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t","coin":195,"type":"TRC20"},{"name":"Wuhan Fried Bats","symbol":"WUHAN","decimals":4,"token_id":"TSzjFRf8bQ46kRndWaHYKSq1P5HzRJfPvr","coin":195,"type":"TRC20"},{"name":"HelGro","symbol":"HGRO","decimals":6,"token_id":"TTvVC9jv5AfDdHuGeCMcQg6QftmNnfQiVm","coin":195,"type":"TRC20"}]`
```
Comment (vikmeup) [inline platform/tron/token_test.go:105]: Move this to files

File: api/registry.go (hunk 1 above)
Comment (EnoRage) [reply api/registry.go:119]: It could be, no critical difference

File: db/asset.go (hunk 2 above)
Comment (EnoRage) [reply db/asset.go:57]: will use memory cache

File: pkg/blockatlas/tx.go (hunk 4 above)
Comment (EnoRage) [reply pkg/blockatlas/tx.go:461]: Good idea, it is possible to make a common function to parse data like from, to,symbol and etc and re-use in 1-3 places

File: platform/tron/token_test.go (hunk 5 above)
Comment (EnoRage) [reply platform/tron/token_test.go:105]: https://github.com/trustwallet/blockatlas/issues/1250

File: db/asset.go (hunk 2 above)
Comment (EnoRage) [reply db/asset.go:57]: https://github.com/trustwallet/blockatlas/issues/1251

**Review by maxUo — APPROVED** [review]

**Review by EnoRage — APPROVED** [review]

**Pull request merged**